toolchain/check/import_ref.cpp

```diff
@@ -409,42 +437,154 @@ class ImportRefResolver {
+    }
+    return GetLocalConstantId(
+        import_ir_.insts()
+            .Get(import_ir_.generics().Get(generic_id).decl_id)
+            .type_id());
+  }
+
+  // Gets a local generic ID given the corresponding local constant ID returned
+  // by GetLocalConstantId for the imported generic. Does not add any new work.
+  auto GetLocalGenericId(SemIR::ConstantId local_const_id) -> SemIR::GenericId {
+    if (!local_const_id.is_valid()) {
+      return SemIR::GenericId::Invalid;
+    }
+    auto type = context_.insts().Get(
+        context_.constant_values().GetInstId(local_const_id));
+    CARBON_KIND_SWITCH(type) {
+      case CARBON_KIND(SemIR::FunctionType fn_type): {
+        return context_.functions().Get(fn_type.function_id).generic_id;
+      }
+      case CARBON_KIND(SemIR::GenericClassType class_type): {
+        return context_.classes().Get(class_type.class_id).generic_id;
+      }
+      case CARBON_KIND(SemIR::GenericInterfaceType interface_type): {
+        return context_.interfaces()
+            .Get(interface_type.interface_id)
+            .generic_id;
+      }
+      default: {
+        CARBON_FATAL() << "Unexpected type for generic declaration: " << type;
+      }
+    }
+  }
+
+  // Gets local information about an imported specific.
+  auto GetLocalSpecificData(SemIR::SpecificId specific_id) -> SpecificData {
     if (!specific_id.is_valid()) {
-      return SemIR::SpecificId::Invalid;
+      return {.generic_const_id = SemIR::ConstantId::Invalid, .args = {}};
     }
     const auto& specific = import_ir_.specifics().Get(specific_id);
-    // TODO: Import the generic.
-    auto generic_id = SemIR::GenericId::Invalid;
-    auto args_id = GetLocalCanonicalInstBlockId(specific.args_id, args);
-    // TODO: Also import the specific.
+    return {
+        .generic_const_id = GetLocalConstantId(specific.generic_id),
+        .args = GetLocalInstBlockContents(specific.args_id),
+        .decl_block = GetLocalInstBlockContents(specific.decl_block_id),
+        .definition_block =
+            GetLocalInstBlockContents(specific.definition_block_id),
+    };
+  }
+
+  // Gets a local specific whose data was already imported by
+  // GetLocalSpecificData. Does not add any new work.
+  auto GetLocalSpecific(SemIR::SpecificId import_specific_id,
+                        const SpecificData& data) -> SemIR::SpecificId {
+    if (!import_specific_id.is_valid()) {
+      return SemIR::SpecificId::Invalid;
+    }
+
+    // Form a corresponding local specific ID.
+    const auto& import_specific =
+        import_ir_.specifics().Get(import_specific_id);
+    auto generic_id = GetLocalGenericId(data.generic_const_id);
+    auto args_id =
+        GetLocalCanonicalInstBlockId(import_specific.args_id, data.args);
+
+    // Populate the specific. Note that we might get data from multiple
+    // different import IRs, so only import data we don't already have.
+    auto specific_id = context_.specifics().GetOrAdd(generic_id, args_id);
```

Comment:
Should this specific_id be returned instead of calling GetOrAdd again below? If not, can you add some comments explaining what's going on?

Separately, if specific_id is pre-existing, is it still necessary to do the sets below? i.e., are the `is_valid` checks mainly because `GetOrAdd` doesn't include a signal for whether the import was already done, or because of forward declarations being an issue? (particularly `decl_block_id` being conditional)

Reply:
Changed to return `specific_id` (just an oversight through refactoring I think).

Setting the definition region conditionally is necessary because we might import the same specific from two IRs, where one of them has a definition and the other does not. The declaration region could be set only the first time, but I think it's reasonable to be doing the check like this, given that it mirrors what we do for the definition.

(I think we might also want to form specifics without even resolving the declaration region at some point. But for now we always resolve the declaration immediately when forming the specific. Conversely, I think we might want to consider lazily importing the specific regions at some point.)